src/node.rs

```diff
@@ -99,6 +99,7 @@ impl<S: StorageExt> SawtoothRaftNode<S> {
     }
 
     pub fn on_block_new(&mut self, block: Block) {
+        let mut leader_state_publishing_this_block = false;
```

Comment:
Instead of making this mutable and setting it later, you can combine some of this code:
`let leader_state_publishing_this_block = match self.leader_state { ... };`
`if leader_state_publishing_this_block { ... }`
`if leader_state_publishing_this_block || self.follower_state.is_some() { ... }`

Reply:
This could be done. Do you see any problem in using mutable variable. Or do you mean the approach you suggested above is more idiomatic in rust.
This fix was done in order to avoid a case where a follower becomes a leader (Node 1) and previous leader (Node 2) had sent a new block just before changing to follower. Node 2 received this block later. At same time Node 2 also has started building a block on same head. So we'll get a case where Node will have 2 blocks added to queue with same block number but different block_id , both pointing to same previous head. Leader node should accept only those blocks published by the leader

Reply:
So the issue is that a leader node publishes a new block before stepping down, and the new leader also publishes a block at the same height. Is that correct? If so, then I think the better way to solve this problem is to not publish (finalize) a block until after the node's state is updated. This should be pretty simple to accomplish by moving the call to `node.check_publish` out of the `node.tick` method and instead call it from `engine.rs` after the call to `process_ready`.

Reply:
That's right. I think its good for leader node to accept only the blocks it publishes. So making this explicit check in code is better. Even if we rely on order of calling '''node.check_publish''' after '''process_ready''',  can the new block can arrive other nodes at any time based on the network delays ? What if block was published and then leader changed to follower in next round and block took long time to reach another node which became leader.

Reply:
I agree that implementing checks for received blocks is a good idea. But I think there should probably be a broader fix for this issue. There are various situations where multiple blocks could be received at the same block height: network delays, bad-actors (e.g. some random follower publishes a block), etc. Also, part of the consensus engine's job is to decide on **all** blocks it receives, to either accept or reject them. So if the node gets a bad block, it needs to tell the validator to reject it. These things should all be considered as a complete solution.

Reply:
See the "API contracts" section of the consensus API RFC: https://github.com/hyperledger/sawtooth-rfcs/blob/master/text/0004-consensus-api.md

Reply:
I thought we are not solving for all bad actor cases as Raft is not BFT though we can still consider cases where it can be handled with minimal checks. I've considered only case where block is received at same block height due to network delays. Do you mean we should also call fail_block() when leader receives block at same height not published by it (to avoid any kind of memory leaks in validator).  And I guess same should be done at follower end when it receives updated log entry from leader. So if a follower commits a log entry, fail all blocks which is not yet committed and whose height is less than recent committed block ?

Reply:
Updated the commit by calling fail_block() if the new block is not handled